File: src/gubg/tui.cpp
#include <gubg/mss.hpp>
#include <gubg/tui.hpp>

#include <sys/ioctl.h>
#include <termios.h>
#include <unistd.h>

namespace gubg {
namespace tui {

bool get_size(unsigned int &width, unsigned int &height)
{
  MSS_BEGIN(bool);

  struct winsize sz{};
  MSS(::ioctl(0, TIOCGWINSZ, &sz) == 0);

  width = sz.ws_col;
  height = sz.ws_row;

  MSS_END();
}

bool get_char(char &ch) {
  MSS_BEGIN(bool);

  struct termios old = {0};
  MSS(::tcgetattr(0, &old) == 0);

  old.c_lflag &= ~ICANON;
  old.c_lflag &= ~ECHO;
  old.c_cc[VMIN] = 1;
  old.c_cc[VTIME] = 0;

  MSS(::tcsetattr(0, TCSANOW, &old) == 0);

  MSS(read(0, &ch, 1) > 0);

  old.c_lflag |= ICANON;
  old.c_lflag |= ECHO;
  MSS(::tcsetattr(0, TCSADRAIN, &old) == 0);

  MSS_END();
}

} // namespace tui
} // namespace gubg

File: src/gubg/tui.zig
const std = @import("std");

pub const Error = error{
    Unknown,
    TCGetAttr,
    TCSetAttr,
    IOCtl,
};

pub const Terminal = struct {
    const Self = @This();
    const platform = std.os.linux;

    termios: std.c.termios = undefined,

    const Config = struct {
        timeout: bool = false,
    };
    pub fn init(config: Config) Error!Terminal {
        var self = Terminal{};

        if (std.c.tcgetattr(0, &self.termios) != 0)
            return Error.TCGetAttr;

        var my = self.termios;

        // Canonical mode: do not wait for <enter>
        my.lflag &= ~platform.ICANON;
        // Do not echo each typed character
        my.lflag &= ~platform.ECHO;

        if (config.timeout) {
            my.cc[platform.V.TIME] = 1;
            my.cc[platform.V.MIN] = 0;
        } else {
            my.cc[platform.V.TIME] = 0;
            my.cc[platform.V.MIN] = 1;
        }

        if (std.c.tcsetattr(0, platform.TCSA.NOW, &my) != 0)
            return Error.TCSetAttr;

        return self;
    }

    pub fn deinit(self: *Self) void {
        _ = std.c.tcsetattr(0, platform.TCSA.NOW, &self.termios);
    }

    const CharSize = struct {
        width: u32 = 0,
        height: u32 = 0,
    };
    pub fn get_charsize() Error!CharSize {
        var sz: platform.winsize = undefined;

        if (std.c.ioctl(0, platform.T.IOCGWINSZ, &sz) != 0)
            return Error.IOCtl;

        return CharSize{
            .width = sz.ws_col,
            .height = sz.ws_row,
        };
    }

    pub fn read_char(_: *Self) ?u8 {
        var buf: [1]u8 = undefined;
        if (std.c.read(0, &buf, buf.len) == 1)
            return buf[0];
        return null;
    }
};
